[WIP] Introduce HeatSources

masterleinad opened the pull request:
This is a step towards https://github.com/adamantine-sim/adamantine/issues/190. This pull request replaces `std::vector<std::shared_ptr<adamantine::HeatSource<dim>>>` with `HeatSources` that contains all heat sources as Kokkos::Views and loops over them in its member functions. With this approach, there is no dynamic polymorphism and the compiler could inline all function calls for the heat sources. In particular, this object would be usable on GPUs.

The tests are still failing but I wanted to put this out already to see if there are objections to this approach.

stvdwtt commented on source/ScanPath.cc:
```diff
@@ -138,7 +140,9 @@ void ScanPath::load_event_series_scan_path(std::string scan_path_file)
 
-    _segment_list.push_back(segment);
+    if (_segment_list_length == MAX_NUMBER_OF_SEGMENTS)
+      Kokkos::abort("too many segmnets");
```
Typo:
```suggestion
      Kokkos::abort("too many segments");
```

stvdwtt commented on source/ScanPath.hh:
```diff
@@ -19,6 +19,8 @@
 #include <istream>
 #include <vector>
 
+#define MAX_NUMBER_OF_SEGMENTS 100
```
This is too low -- if we had a 50-layer print with only two segments per layer, we'd hit this. I'm not sure that we've done enough sliced full parts to know a good upper bound. For example the AOP hourglass has 8055 segments and that's pretty simple as far as "real" parts go. Can we use a Kokkos::View with a runtime-defined length instead?

masterleinad replied:
That was one of my next steps but I would need to manage the allocations for the segments outside the heat sources to avoid Views of managed Views. I'm not sure how much freedom we have to change those APIs.

The current version should be (more or less) device-usable already at least.

stvdwtt replied:
Ok, got it. Maybe another option would be to make it a template parameter? Even with the dynamic scan path in #223, we wouldn't be changing the length of the scan path too many times per simulation.

masterleinad replied:
Sure, changing it to a template argument for the heat source and the scan path is easy. The question then is if we want to have a runtime dispatch for pre-instantiated maximum sizes or any other mechanism that doesn't require rebuilding the library and the application for certain input files.

Rombur replied:
We don't want to do this. We don't know how many segments we will need but it's potentially very large. If someone wants to point-source raster for PBF, the number of segments will be huge. 
In the current code, the scan path is part of the heat source. You need to change that. Move the scan path to `HeatSources`, add a function that evaluates the scan patch from `t_0` to `t_1` and fill a `Kokkos::View`. This way we don't need to know the number of segment and you don't waste memory if you have a ton a segments.

> I'm not sure how much freedom we have to change those APIs.

`adamantine` is not a library. Everything but the input parameters is an implementation details.

Rombur commented on source/HeatSources.hh:
```diff
@@ -0,0 +1,404 @@
+  return value;
+}
+
+template <typename MemorySpace, int dim>
+std::vector<ScanPath<MemorySpace>>
+HeatSources<MemorySpace, dim>::get_scan_paths() const
+{
+  std::vector<ScanPath<MemorySpace>> scan_paths;
+  for (unsigned int i = 0; i < _electron_beam_heat_sources.size(); ++i)
+    scan_paths.push_back(_electron_beam_heat_sources(i).get_scan_path());
+  for (unsigned int i = 0; i < _goldak_heat_sources.size(); ++i)
+    scan_paths.push_back(_goldak_heat_sources(i).get_scan_path());
+  return scan_paths;
+}
```
No it's not. There is no order in the input file. You can reorder everything and the same input file will work.

Rombur commented on source/ThermalPhysics.templates.hh:
```diff
@@ -844,21 +809,9 @@ void ThermalPhysics<
-
-    source_index++;
-  }
+  auto host_heat_sources = _heat_sources.copy_to(dealii::MemorySpace::Host{});
```
```suggestion
  auto heat_sources_host = _heat_sources.copy_to(dealii::MemorySpace::Host{});
```

Rombur commented on source/ThermalPhysics.templates.hh:
```diff
@@ -844,21 +809,9 @@ void ThermalPhysics<
-    source_index++;
-  }
+  auto host_heat_sources = _heat_sources.copy_to(dealii::MemorySpace::Host{});
+  host_heat_sources.set_beam_properties(heat_source_database);
```
We know that the database will always live on the host. It seems wasteful to copy everything back and forth.

Rombur commented on source/ThermalPhysics.templates.hh:
```diff
@@ -870,15 +823,8 @@ double ThermalPhysics<dim, p_order, fe_degree, MaterialStates, MemorySpaceType,
-  }
-  _current_source_height = temp_height;
+  _current_source_height =
+      _heat_sources.copy_to(dealii::MemorySpace::Host{}).get_current_height(t);
```
Same comment as above

Rombur commented on source/ThermalPhysicsInterface.hh:
```diff
@@ -19,9 +19,6 @@ namespace adamantine
 
-template <int dim>
-class HeatSource;
-
```
You don't need a forward declaration because you are relying on the order of the includes.

Rombur replied:
You still need a forward declaration, otherwise the code may break if the include are reordered.

Rombur commented on source/ScanPath.hh:
```diff
@@ -66,13 +69,16 @@ public:
+  }
+
+  static std::vector<ScanPathSegment>
+  extract_scan_paths(std::string scan_path_file, std::string file_format);
```
I don't really like this name. You are just reading from a file.

Rombur commented on tests/test_heat_source.cc:
```diff
@@ -29,8 +29,17 @@ BOOST_AUTO_TEST_CASE(heat_source_value_2d, *utf::tolerance(1e-12))
+  GoldakHeatSource<2, dealii::MemorySpace::Host> goldak_heat_source(
+      beam, ScanPath<dealii::MemorySpace::Host>(scan_paths_segments_view));
+  ElectronBeamHeatSource<2, dealii::MemorySpace::Host> eb_heat_source(
+      database, ScanPath<dealii::MemorySpace::Host>(scan_paths_segments_view));
```
Can you put this code in a function instead of copy/pasting three times.

Rombur commented on tests/test_material_deposition.cc:
```diff
@@ -353,7 +353,14 @@ BOOST_AUTO_TEST_CASE(material_deposition)
+      scan_paths_segments_view(scan_path_segments.data(),
+                               scan_path_segments.size());
+  adamantine::ScanPath<dealii::MemorySpace::Host> scan_path(
+      scan_paths_segments_view);
```
Put this code in a function to avoid the copy/paste.

Rombur reviewed (changes requested):
You will need to rebase the PR on develop. There was some small changes in `ScanPath`.

masterleinad commented:
Retest this please.

Rombur commented:
Let me know when you want another review on this.

masterleinad commented:
> Let me know when you want another review on this.

This could take another review now. Some notes:
- I create local copies on the host for the HeatSources object where necessary assuming that we want it by default to be in the memory space specified in the respective template argument for the other operators. It might make sense to keep it in host memory instead and only create device copies or we keep a member variable for both memory spaces (if they don't coincide).
- I needed to move a bunch of `const` members in the heat sources to make these classes copyable. I was running into some issues with `static` variables if used on the device as well. Thus, I moved the respective constants into member functions.
- The recent changes reusing scan file paths made the HeatSources code a little more complicated. We are now storing those `std::strings` in `HeatSources` directly as well and recreating the scan file segments from them if necessary. Since the `ScanPath` doesn't own the `ScanPathSegments`, reading the file is a static member instead of a member function named "extract_scan_paths". I'm happy to change it back to some other name like "read_file" again if preferred.
- I believe I addressed all other comments.

Rombur commented on source/ElectronBeamHeatSource.hh:
```diff
@@ -33,25 +34,82 @@ public:
+  void set_scan_path(ScanPath<MemorySpaceType> const scan_path)
+  {
+    _scan_path = scan_path;
+  }
```
Move the implementation out of the class declaration

Rombur commented on source/ElectronBeamHeatSource.hh:
```diff
@@ -33,25 +34,82 @@ public:
+  }
+
+  /**
+   * Compute the current height of the where the heat source meets the material
```
```suggestion
   * Compute the current height where the heat source meets the material
```

Rombur commented on source/GoldakHeatSource.cc:
```diff
@@ -9,29 +9,36 @@
+void GoldakHeatSource<dim, MemorySpaceType>::update_time(double time)
 {
+  static const double pi_over_3_to_1p5 =
+      std::pow(dealii::numbers::PI / 3.0, 1.5);
```
Why the change?

Rombur commented on source/GoldakHeatSource.hh:
```diff
@@ -33,25 +34,82 @@ public:
+  void set_scan_path(ScanPath<MemorySpaceType> const scan_path)
+  {
+    _scan_path = scan_path;
+  }
```
Move the implementation out of the declaration

Rombur commented on source/HeatSources.hh:
```diff
@@ -0,0 +1,482 @@
+   */
+  template <typename Dummy = MemorySpaceType>
+  std::enable_if_t<std::is_same_v<Dummy, dealii::MemorySpace::Host>>
+  update_scan_paths();
```
`update_scan_paths()` should just do the right thing independent of the `MemorySpace`

Rombur commented on source/HeatSources.hh:
```diff
@@ -0,0 +1,482 @@
+  void internal_update_scan_paths();
+
+  /**
+   * Private constructor used by copy_to_host.
```
I cannot find `copy_to_host`

Rombur commented on source/HeatSources.hh:
```diff
@@ -0,0 +1,482 @@
+
+template <int dim, typename MemorySpaceType>
+void HeatSources<dim, MemorySpaceType>::internal_update_scan_paths()
+{
```
Put the implementation in a `.cc` not in the header.

Rombur commented on application/adamantine.hh:
```diff
@@ -1025,25 +1032,26 @@ run(MPI_Comm const &communicator, boost::property_tree::ptree const &database,
+        heat_sources_host.update_scan_paths();
+        heat_sources = heat_sources_host.copy_to(MemorySpaceType{});
+
+        auto const &scan_paths = heat_sources.get_scan_paths();
```
This looks wrong. Shouldn't this be `heat_source_host.get_scan_path()`?

Rombur commented on source/HeatSources.hh:
```diff
@@ -0,0 +1,482 @@
+
+  for (unsigned int i = 0; i < _goldak_heat_sources.size(); ++i)
+    _goldak_heat_sources(i).set_scan_path(
+        ScanPath<MemorySpaceType>{_goldak_scan_path_segments[i]});
```
Why are we keeping track of the scan path in `HeatSources` and in each individual sources? We should not duplicate this information.

Rombur commented on source/HeatSources.hh:
```diff
@@ -0,0 +1,455 @@
+    value += _cube_heat_sources(i).value(point, height);
+  for (unsigned int i = 0; i < _goldak_heat_sources.size(); ++i)
+    value += _goldak_heat_sources(i).value(point, height);
+  return value;
```
?

Rombur commented on source/ScanPath.cc:
```diff
@@ -8,41 +8,45 @@
+
+template <typename MemorySpaceType>
+std::vector<ScanPathSegment>
+ScanPath<MemorySpaceType>::extract_scan_paths(std::string scan_path_file,
```
I still hate the name

Rombur commented on source/ScanPath.hh:
```diff
@@ -54,6 +54,7 @@ class ScanPathTester;
  * gives the power modifier for the current segment. It reads in the scan path
  * from a text file.
  */
+template <typename MemorySpaceType>
```
Is is actually necessary to have the ScanPath on the GPU? I feel like this should only exists on the CPU and we should only move part of the data when it's necessary. Otherwise, we will end up using several GBs of memory on the GPU just for the ScanPath.
